### PR TITLE
Remove arbitrary props from DOM elements

### PR DESCRIPTION
**Pull request by AugustinLF**
This is a work in progress, since I'm stuck with a couple of case. Perhaps someone could help me with that? I skipped the two failing tests, I'll comment to explain my difficulties.

### CODE REVIEW

File: test/ReactWrapper-spec.js [hunk 1]
```diff
@@ -372,7 +372,7 @@ describeWithDOM('mount', () => {
 
     });
 
-    it('should not find components with invalid attributes', () => {
+    it.skip('should not find components with invalid attributes', () => {
```
Comment (AugustinLF) [inline test/ReactWrapper-spec.js]: The point of this test is to test on purpose wrong attributes that should be ignored by enzyme. I'm not sure there's anything to do. Since I don't like warnings when running tests, I would do:
```
const _warn = console.warn;
console.warn = () => {};
// tests
console.warn = _warn;
```

File: test/ReactWrapper-spec.js [hunk 2]
```diff
@@ -502,16 +502,21 @@ describeWithDOM('mount', () => {
...
-          <button data-test="ref" prop={undefined} />
-          <span data-test="ref" prop={null} />
-          <div data-test="ref" prop={123} />
-          <input data-test="ref" prop={false} />
-          <a data-test="ref" prop />
+          <Foo prop={undefined} />
+          <Foo prop={null} />
+          <Foo prop={123} />
+          <Foo prop={false} />
+          <Foo prop />
```
Comment (AugustinLF) [inline test/ReactWrapper-spec.js]: Here, all the associated find tests `find({ prop: undefined })` fail when using a React Component instead of a simple element. I'm quite surprised of this behaviour, for me it should behave in the same way, perhaps there is a enzyme behaviour that I'm unaware of?

Comment (Aweary) [inline test/ReactWrapper-spec.js]: That will likely be resolved by https://github.com/airbnb/enzyme/pull/377. Why did change this from using DOM nodes to a component, though? I think we should keep this test as-is.

Comment (ljharb) [inline test/ReactWrapper-spec.js]: Or better, keep the DOM tests and still add the Foo tests.

File: test/ReactWrapper-spec.js (hunk 1 above)
Comment (Aweary) [inline test/ReactWrapper-spec.js]: We could probably just use `sinon.spy` to mock `console.warn` in the test itself, and revert the mock (does sinon do that by itself? I'm not sure...) in the `beforeEach` block.

File: test/ReactWrapper-spec.js (hunk 2 above)
Comment (Aweary) [inline test/ReactWrapper-spec.js]: Since it fails now with components, I can just those tests to #377.

File: test/ReactWrapper-spec.js (hunk 1 above)
Comment (ljharb) [inline test/ReactWrapper-spec.js]: sinon-sandbox will do that by itself - if we're not using that yet (and restoring it afterEach test), we should

**Comment by AugustinLF** [comment]
I updated the PR to fix the omitted tests, and rewrite some using `data-amount`, `data-foo`, to follow @ljharb. If everything's fine, I'll squash the commits, and rebase the whole branch!

**Comment by Aweary** [comment]
LGTM @AugustinLF can you resolve the merge conflicts here? I think we can merge afterwards.

**Comment by blainekasten** [comment]
closes #480

**Comment by AugustinLF** [comment]
Done! Should I squash the commits?

File: test/ReactWrapper-spec.jsx [hunk 3]
```diff
@@ -385,6 +394,8 @@ describeWithDOM('mount', () => {
     });
 
     it('should not find components with invalid attributes', () => {
+      // React 15.2 warns when setting a non valid prop to an DOM element
+      sandbox.stub(console, 'error');
```
Comment (ljharb) [inline test/ReactWrapper-spec.jsx]: why is this necessary? the test will fail if a non-valid prop is set, won't it?

Comment (AugustinLF) [inline test/ReactWrapper-spec.jsx]: The test will pass, it will simply warn in the console.
image

Comment (ljharb) [inline test/ReactWrapper-spec.jsx]: If I haven't already, I'll soon be making console warnings and errors fail tests - so if a test would be issuing a warning in a version of React, we shouldn't be running that test in that version of React (since React covers it for us by issuing the warning).

Can we restrict this test to only React < 15, and remove the stub?

Comment (AugustinLF) [inline test/ReactWrapper-spec.jsx]: Done! I changed so that the test is only run when react 0.13 or 0.14.

File: package.json [hunk 4]
```diff
@@ -84,6 +84,6 @@
     "sinon": "^1.17.5"
   },
   "peerDependencies": {
-    "react": "0.13.x || 0.14.x || ^15.0.0-0"
+    "react": "0.13.x || 0.14.x || ^15.x"
```
Comment (ljharb) [inline package.json]: Is this change necessary? We want to include the v15 prereleases.

Comment (AugustinLF) [inline package.json]: No it's not anymore. I did it because I wrote that line before the release of 15.2. I was working with 15.2.0-rc.1
 or 2, which doesn't match ^15.0.0-0). I added a commit to fix that.

Comment (ljharb) [inline package.json]: ah, good call. Let's make this `0.13.x || 0.14.x || ^15.0.0-0 || 15.x` then :-)

Comment (AugustinLF) [inline package.json]: Let's hope it's fine now!

**Comment by ljharb** [comment]
@AugustinLF i think you can simply remove the last two commits now :-)

**Comment by AugustinLF** [comment]
@ljharb Yep, it's done!

**Comment by AugustinLF** [comment]
I rebased the branch to fix the conflicts :)

**Comment by Aweary** [comment]
Thanks @AugustinLF!
